jax/core.py

```diff
@@ -1102,6 +1102,39 @@ def shape(self):
            "UnshapedArray instances to ever be produced.")
     raise TypeError(msg)
 
+class DShapedArray(UnshapedArray):
+  __slots__ = ['shape']
+  array_abstraction_level = 2
+
+  def __init__(self, shape, dtype, weak_type):
+    self.shape = shape
```

Comment:
Is it worth commenting, possibly in the form of a type annotation, the set of values that may appear in a shape?

Previously it was trivial ("int"s), but now it isn't and it's not obvious.

Reply:
Great idea. Done. I took your suggestion to document it as a type annotation, and I added this:

```python
# We have a convention of reusing AbsractValues as types, in particular reusing
# ShapedArrays as types, even though we could make a distinction and use
# abstract values during tracing only. This reuse becomes a bit more extreme
# with DShapedArrays. A DShapedArray's shape attribute is a tuple which can
# contain several different types: ints, other AbstractValues (specifically at
# the input and output to pe.trace_to_jaxpr_dynamic), Tracers (while tracing),
# or Vars (when used as jaxpr type annotations). We could reduce this
# polymorphism if it seems cleaner, though it's kind of convenient!
AxisSizeForTracing = Union[int, Tracer]
AxisSizeForJaxprType = Union[int, Var]
AxisSizeForJaxprTracingSpec = Union[int, AbstractValue]
AxisSize = Union[AxisSizeForTracing, AxisSizeForJaxprType,
                 AxisSizeForJaxprTracingSpec]

class DShapedArray(UnshapedArray):
  __slots__ = ['shape']
  shape: Tuple[AxisSize, ...]  # see comment above
  array_abstraction_level = 2
  ...
```